packages/reporters/dev-server/src/Server.js

```diff
@@ -0,0 +1,204 @@
+  );
+  res.setHeader(
+    'Access-Control-Allow-Headers',
+    'Origin, X-Requested-With, Content-Type, Accept, Content-Type'
+  );
+}
+
+export default class Server extends EventEmitter {
+  pending: boolean;
+  options: DevServerOptions;
+  serve: (
+    req: Request,
+    res: Response,
+    next?: (req: Request, res: Response, next?: (any) => any) => any
+  ) => any;
+  bundleGraph: BundleGraph | null;
+  error: PrintableError | null;
+
+  constructor(options: DevServerOptions) {
+    super();
+
+    this.options = options;
+    this.pending = true;
+    this.bundleGraph = null;
+    this.error = null;
+
+    this.serve = serveStatic(this.options.distDir, {
+      index: false,
+      redirect: false,
+      setHeaders: setHeaders,
+      dotfiles: 'allow'
+    });
+  }
+
+  buildSuccess(bundleGraph: BundleGraph) {
+    this.bundleGraph = bundleGraph;
+    this.pending = false;
+
+    this.emit('bundled');
+  }
+
+  buildError(error: PrintableError) {
+    this.error = error;
+  }
+
+  respond(req: Request, res: Response) {
+    let {pathname} = url.parse(req.url);
+
+    if (this.error) {
+      return this.send500(req, res);
+    } else if (
+      !pathname ||
+      !pathname.startsWith(this.options.publicUrl) ||
+      path.extname(pathname) === ''
+    ) {
+      // If the URL doesn't start with the public path, or the URL doesn't
+      // have a file extension, send the main HTML bundle.
+      return this.sendIndex(req, res);
+    } else {
+      // Otherwise, serve the file from the dist folder
+      req.url = pathname.slice(this.options.publicUrl.length);
+      return this.serve(req, res, () => this.sendIndex(req, res));
+    }
+  }
+
+  async sendIndex(req: Request, res: Response) {
+    if (this.bundleGraph) {
+      // If the main asset is an HTML file, serve it
+      let htmlBundle = null;
+      this.bundleGraph.traverseBundles(bundle => {
+        if (bundle.type === 'html' && bundle.isEntry) {
+          htmlBundle = bundle;
+        }
+      });
+
+      if (htmlBundle) {
+        req.url = `/${path.basename(htmlBundle.filePath)}`;
+
+        this.serve(req, res, () => this.send404(req, res));
+      } else {
+        this.send404(req, res);
+      }
+    } else {
+      this.send404(req, res);
+    }
+  }
+
+  async send404(req: Request, res: Response) {
+    res.statusCode = 404;
+
+    let template404 = (await readFile(
+      path.join(__dirname, 'templates/404.html')
+    )).toString('utf8');
+
+    setHeaders(res);
+    res.end(template404);
+  }
+
+  async send500(req: Request, res: Response) {
+    setHeaders(res);
+
+    res.setHeader('Content-Type', 'text/html; charset=utf-8');
+    res.writeHead(500);
+
+    if (this.error) {
+      const {message, stack} = prettyError(this.error, {color: true});
+      let stackHTML = ansiHtml(stack);
+      let template500 = (await readFile(
+        path.join(__dirname, 'templates/500.html')
+      )).toString('utf8');
+
+      res.end(
+        template500
```

Comment:
Maybe... use a real templatng language? 😉

Reply:
Originally was gonna use pug but wasn't sure if I should add another dep. But will do makes sense to do it...